lib/_http_client.js

```diff
@@ -362,52 +348,55 @@ ClientRequest.prototype.abort = function abort() {
   }
   this.aborted = true;
   process.nextTick(emitAbortNT, this);
+
+  // Compat. If we're aborting, we don't care about any more response data.
+  this.res?._dump();
+
   this.destroy();
 };
 
-ClientRequest.prototype.destroy = function destroy(err) {
-  if (this.destroyed) {
-    return this;
+ClientRequest.prototype._construct = function(callback) {
+  if (this.socket) {
+    callback();
+  } else {
+    this.once(kConstructed, callback);
   }
-  this.destroyed = true;
+};
 
-  // If we're aborting, we don't care about any more response data.
-  if (this.res) {
-    this.res._dump();
+ClientRequest.prototype._destroy = function(err, callback) {
+  if (!err && !this.aborted && !this.writableEnded) {
+    err = connResetException('socket hang up');
+  }
+
+  // TODO (ronag): Check if socket was used at all (e.g. headersSent) and
+  // re-use it in that case. `req.socket` just checks whether the socket was
+  // assigned to the request and *might* have been used.
+
+  if (this.res && !this.res.complete) {
+    this.res.destroy(connResetException('aborted'));
   }
 
   // In the event that we don't have a socket, we will pop out of
   // the request queue through handling in onSocket.
   if (this.socket) {
-    _destroy(this, this.socket, err);
-  } else if (err) {
-    this[kError] = err;
+    if (err) {
+      // Compat
+      this.socket._hadError = true;
+    }
+
+    this.socket.destroy(err);
+    finished(this.socket, (er) =>
+      // XXX: This needs more thought.
+      callback(this.listenerCount('error') === 0 ?
```

Comment:
Just trying to help to sort out tests. 

`test/parallel/test-http-client-abort-keep-alive-queued-tcp-socket.js` goes into timeout because [this](https://github.com/nodejs/node/blob/59daacfd2b4e5cdc6750682ade485080b13d374c/test/parallel/test-http-client-abort-keep-alive-queued-tcp-socket.js#L34) never gets called. The `listenerCounts` is `0` in that case. I am trying to understand why, given that a handler is registered one line before.

Reply:
And of course, without that logic, an unhandled `error` event is emitted.